src/AbstractValueParser.php

```diff
@@ -28,7 +41,9 @@ public function defaultsTo($defaultValue)
 
     public function required()
     {
-        if (is_null($this->value)) {
+        if ($this->invalid) {
+            throw $this->createInvalidValueException();
+        } elseif (is_null($this->value)) {
```

Comment:
Maybe I'm missing something, but if we construct the class and set `invalid` to `true` when `value === null`, then `required()` is sending a mixed message because it throws `$this->createInvalidValueException()` when the value is invalid, but also throws `$this->createNotFoundException()` if `is_null`. Is that the expected behaviour?